### DISCUSSION TITLE
Sanitization problems after Chem.PathToSubmol

### DISCUSSION THREAD

**Discussion by jepdavidson** [rdkit/rdkit#6481] [Q&A] [answered]
Dear All,

I am looking for a bit of help/advice on the 'proper' way to clean-up structures returned from calling Chem.PathToSubmol (rdkit 2023.03.2).

Here's a convenience function to return a submol from a smiles string, given a smarts query:

```python
from rdkit import Chem, rdBase
from rdkit.Chem import Draw

print(rdBase.rdkitVersion)


def get_submol(smi: str, smarts: str) -> Chem.Mol:
    mol = Chem.MolFromSmiles(smi)
    query = Chem.MolFromSmarts(smarts)
    # Get the bond idxs
    atom_idxs = mol.GetSubstructMatch(query)
    bond_idxs = [
        b.GetIdx() for b in mol.GetBonds() if 
            set([b.GetBeginAtomIdx(), b.GetEndAtomIdx()]).issubset(set(atom_idxs))
    ]
    return Chem.PathToSubmol(mol, bond_idxs)
```
> 2023.03.2

Here are some examples to work with:

```python
s1 = "COCCC(C)NC(C)=O"
s2 = "COCC[C@H](C)NC(C)=O"
s3 = "COCC[C@@](C)(F)NC(C)=O"
Draw.MolsToGridImage([Chem.MolFromSmiles(smi) for smi in (s1, s2, s3)])
```
![image](https://github.com/rdkit/rdkit/assets/7816101/0703a4d9-f436-4a42-8d18-304713aba6b5)

And here are the extracted submols if we extract the methyl acetamide portion from each:

```python
query = "CC(=O)NC"
submols = [get_submol(smi, query) for smi in (s1, s2, s3)]
Draw.MolsToGridImage(submols, legends=[Chem.MolToSmiles(m) for m in submols])
```
![image](https://github.com/rdkit/rdkit/assets/7816101/f6771b9f-60f6-4a95-ac71-58d1fd66828f)

The stereo information appears to still be present.  Not a problem - probably `PathToSubmol` isn't sanitizing things(?).

### Chem.SanitizeMol

```python
[Chem.SanitizeMol(m) for m in submols]
Draw.MolsToGridImage(submols, legends=[Chem.MolToSmiles(m) for m in submols])
```
![image](https://github.com/rdkit/rdkit/assets/7816101/f1a8b679-ac61-466d-852d-1bd6c4818b7b)

Hmmm...  We still have the stereo information, but now we have radicals as well(!).

Let's try again, but this time let's try calling Chem.AssignStereochemistery()

### Chem.AssignStereochemistry

```python
submols = [get_submol(smi, query) for smi in (s1, s2, s3)]
[Chem.AssignStereochemistry(m, cleanIt=True, force=True) for m in submols]
Draw.MolsToGridImage(submols, legends=[Chem.MolToSmiles(m) for m in submols])
```
![image](https://github.com/rdkit/rdkit/assets/7816101/0aca99f2-bb90-4508-bd4b-3e24b804a193)

On the face of it these look better, but if we sanitize:

```python
[Chem.SanitizeMol(m) for m in submols]
Draw.MolsToGridImage(submols, legends=[Chem.MolToSmiles(m) for m in submols])
```
![image](https://github.com/rdkit/rdkit/assets/7816101/2b843048-bcf3-4451-8216-687be23dfdae)

Weird(?) - the diradical is 'mended', but the triradical is still there(!).

### Manually changing the number of radical electrons

We *can* fix things like this:

```python
def fix_submol(mol):
    Chem.AssignStereochemistry(mol, cleanIt=True, force=True)
    radical_counts = [atom.GetNumRadicalElectrons() for atom in mol.GetAtoms()]
    Chem.SanitizeMol(mol)
    for atom, count in zip(mol.GetAtoms(), radical_counts):
        atom.SetNumRadicalElectrons(count)

submols = [get_submol(smi, query) for smi in (s1, s2, s3)]
[fix_submol(mol) for mol in submols]
Draw.MolsToGridImage(submols, legends=[Chem.MolToSmiles(m) for m in submols])
```
![image](https://github.com/rdkit/rdkit/assets/7816101/a2106ddc-b341-45e5-956f-549126c63744)

But having to (sometimes) reset the number of radical electrons feels hacky!

It would be great to know the proper way to sanitize the stereochemistry for the extracted submols, as well as understand why the atom degree seems to get corrected by manipulating the number of radical electrons (rather than by perhaps updating the implicit hydrogen count).

As always, any help is greatly appreciated!

Kind regards

James

**Comment by ptosco** [comment]
Hi James, apparently `PathToSubmol()` does not do a great job when disconnecting bonds, i.e., it does not update the implicit H count to fill open valences, so you'll need to do that manually after calling `PathToSubmol()`. A simple way to find out the number of Hs needed to fill empty valences is to let `SanitizeMol()` fill open valences with radical electrons, and then convert those to Hs, and finally recompute stereo.
```
def get_submol(smi: str, smarts: str) -> Chem.Mol:
    mol = Chem.MolFromSmiles(smi)
    query = Chem.MolFromSmarts(smarts)
    # Get the bond idxs
    atom_idxs = mol.GetSubstructMatch(query)
    bond_idxs = [
        b.GetIdx() for b in mol.GetBonds() if 
            set([b.GetBeginAtomIdx(), b.GetEndAtomIdx()]).issubset(set(atom_idxs))
    ]
    res = Chem.PathToSubmol(mol, bond_idxs)
    Chem.SanitizeMol(res)
    for a in res.GetAtoms():
        a.SetNumExplicitHs(a.GetNumExplicitHs() + a.GetNumRadicalElectrons())
        a.SetNumRadicalElectrons(0)
        a.UpdatePropertyCache()
    res.UpdatePropertyCache()
    Chem.AssignStereochemistry(res, True, True)
    return res
```
Side note: your typo
> Let's try again, but this time let's try calling Chem.AssignStereochemistery()

perfectly captures the mysterious essence of stereochemistry: well done!

**Reply by jepdavidson** [reply]
Thanks @ptosco (and thanks for spotting the typo! 😃)

The workaround makes sense, but I guess I am still don't understand why (a) the example with no specified stereochemistry doesn't suffer the same radical electron count increase during sanitization, and (b) why the diradical and triradical behave differently if `AssignStereochemistry` is called before sanitizing...

**Comment by ptosco** [comment, marked as answer]
@jepdavidson 
> I guess I am still don't understand why (a) the example with no specified stereochemistry doesn't suffer the same radical electron count increase during sanitization

Here is the situation for the three molecules before sanitization.
**First molecule**
```
for a in submols[0].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_UNSPECIFIED e 0 i 1 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```
The molecule has one implicit H which already existed in the parent, non-chiral molecule.
Therefore, the sanitizer re-adds the two missing implicit Hs.
After sanitization:
```
0 C CHI_UNSPECIFIED e 0 i 3 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```

**Second molecule**
```
for a in submols[1].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_TETRAHEDRAL_CCW e 1 i 0 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```
This molecule has no implicit Hs, as the parent, chiral molecule had one "explicit" implicit H which was used to determine the stereochemistry.
Therefore, the sanitizer refrains from adding implicit Hs, preserves the pre-existing "explicit" implicit H and fills the two open valences with radicals.
After sanitization:
```
0 C CHI_TETRAHEDRAL_CCW e 1 i 0 r 2
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```

**Third molecule**
```
for a in submols[2].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_TETRAHEDRAL_CW e 0 i 0 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```
This molecule has no implicit Hs, as the parent, chiral molecule has all valences filled by heavy atoms.
Therefore, the sanitizer refrains from adding implicit Hs and fills the three open valences with radicals.
After sanitization:
```
0 C CHI_TETRAHEDRAL_CW e 0 i 0 r 3
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```

<br><br>

> why the diradical and triradical behave differently if `AssignStereochemistry` is called before sanitizing...

Because `AssignStereochemistry` with `force=True` and `cleanIt=True` removes the chiral tags which are not justified by asymmetry anymore after breaking some bonds in the parent molecule.
Therefore, before sanitization you have the following situation:
**First molecule**
```
for a in submols[0].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_UNSPECIFIED e 0 i 1 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```
**Second molecule**
```
for a in submols[1].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_UNSPECIFIED e 0 i 3 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```
In this case the dangling "explicit" implicit H on the carbon, which is not chiral anymore, is removed and implicit Hs recomputed.
**Third molecule**
```
for a in submols[2].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_UNSPECIFIED e 0 i 0 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```
Upon sanitization, the first molecule gets missing implicit Hs re-added as previously.
**First molecule**
```
for a in submols[0].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_UNSPECIFIED e 0 i 3 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```
The second molecule already has all implicit Hs added, so there is no change.
**Second molecule**
```
for a in submols[1].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_UNSPECIFIED e 0 i 3 r 0
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```
For the third molecule, since it has no implicit Hs to start with, the sanitizer refrains from adding implicit Hs and fills its valences with radicals.
**Third molecule**
```
for a in submols[2].GetAtoms():
    print(a.GetIdx(), a.GetSymbol(), a.GetChiralTag(), "e", a.GetNumExplicitHs(), "i", a.GetNumImplicitHs(), "r", a.GetNumRadicalElectrons())
0 C CHI_UNSPECIFIED e 0 i 0 r 3
1 N CHI_UNSPECIFIED e 0 i 1 r 0
2 C CHI_UNSPECIFIED e 0 i 0 r 0
3 C CHI_UNSPECIFIED e 0 i 3 r 0
4 O CHI_UNSPECIFIED e 0 i 0 r 0
```

**Reply by jepdavidson** [reply]
@ptosco - wow, thanks for the very thorough explanation!

I was definitely lacking an appreciation of the "explicit implicit hydrogens"!

Cheers

James

**Comment by ptosco** [comment]
@jepdavidson :-D If you wish to learn more on this fascinating topic, here's some good bedside reading:
https://nextmovesoftware.com/blog/2013/02/27/explicit-and-implicit-hydrogens-taking-liberties-with-valence/
